synapse/storage/data_stores/main/event_federation.py

```diff
@@ -103,6 +103,149 @@ def _get_auth_chain_ids_txn(self, txn, event_ids, include_given, ignore_events):
+        # I.e. if the set is empty then all sets can reach the event.
+        event_to_missing_sets = {
+            event_id: {i for i, a in enumerate(state_sets) if event_id not in a}
+            for event_id in initial_events
+        }
+
+        # We need to get the depth of the initial events for sorting purposes.
+        rows = self.db.simple_select_many_txn(
+            txn,
+            table="events",
+            column="event_id",
+            iterable=initial_events,
+            keyvalues={},
+            retcols=("event_id", "depth"),
+        )
+
+        # The sorted list of events we should walk the auth chain off.
+        search = sorted((row["depth"], row["event_id"]) for row in rows)
+
+        # Map from event to its auth events
+        event_to_auth_events = {}  # type: Dict[str, Set[str]]
+
+        base_sql = """
+            SELECT depth, a.event_id, auth_id
+            FROM event_auth AS a
+            INNER JOIN events AS e ON (e.event_id = a.auth_id)
+            WHERE
+        """
+
+        while search:
+            # Check whether all our current walks are reachable by all state
+            # sets. If so we can bail.
+            if all(not event_to_missing_sets[eid] for _, eid in search):
+                break
+
+            # Fetch the auth events and their depths of the N last events we're
+            # currently walking
+            search, chunk = search[:-100], search[-100:]
+            clause, args = make_in_list_sql_clause(
+                txn.database_engine, "a.event_id", [e_id for _, e_id in chunk]
+            )
+            txn.execute(base_sql + clause, args)
+
+            for depth, event_id, auth_id in txn:
+                event_to_auth_events.setdefault(event_id, set()).add(auth_id)
+
+                if auth_id not in event_to_missing_sets:
+                    # First time we're seeing this event, so we add it to the
+                    # queue of things to fetch.
+                    search.append((depth, auth_id))
+                else:
+                    # We've previously seen this event, so look up its auth
+                    # events and recursively mark all ancestors as reachable
+                    # by the current events state set.
+                    a_ids = event_to_auth_events.get(auth_id)
+                    while a_ids:
+                        new_aids = set()
+                        for a_id in a_ids:
+                            event_to_missing_sets[a_id].intersection_update(
+                                event_to_missing_sets[event_id]
+                            )
+
+                            b = event_to_auth_events.get(a_id)
+                            if b:
+                                new_aids.update(b)
+
+                        a_ids = new_aids
+
+                # Mark that the auth event is reachable by the approriate sets.
+                sets = event_to_missing_sets.setdefault(
```

Comment:
move this to inside the `if` statement at line 219, since it is only relevant if `auth_id` is missing from `event_to_missing_sets`.

Also add a comment:

```
# assume that this event is unreachable from any of the state sets until proven otherwise
```